Pod/Classes/VOKCoreDataManager.m

```diff
@@ -484,78 +495,98 @@ - (NSManagedObjectContext *)safeContext:(NSManagedObjectContext *)context
+        }
+    }];
 }
 
-- (void)saveTempContext:(NSManagedObjectContext *)tempContext
-{
-    [[NSNotificationCenter defaultCenter] addObserver:self
-                                             selector:@selector(tempContextSaved:)
-                                                 name:NSManagedObjectContextDidSaveNotification
-                                               object:tempContext];
+/**
+ Save any changes to a managed object context to the persistent store.
+ If a context has no changes, no saves will be performed.
+ 
+ @param context The context to save
+ @param wait Whether to save synchronously (YES) or asynchronously (NO)
+ */
+- (void)saveContext:(NSManagedObjectContext *)context
+            andWait:(BOOL)wait
+{
+    //The documentation indicates that this method is not thread safe.
+    //We can't guarantee that the current method is called on the appropriate thread for this context,
+    //so use performBlockAndWait to query the context for changes.
+    __block BOOL hasChanges;
+    [context performBlockAndWait:^{
+        hasChanges = context.hasChanges;
+    }];
     
-    [self saveContext:tempContext];
+    if (!hasChanges) {
+        return;
+    }
     
-    [[NSNotificationCenter defaultCenter] removeObserver:self
-                                                    name:NSManagedObjectContextDidSaveNotification
-                                                  object:tempContext];
-}
-
-- (void)tempContextSaved:(NSNotification *)notification
-{
-    // Solved issue with NSFetchedResultsController ignoring changes
-    // merged from different managed object contexts by touching
-    // willAccessValueForKey: on the updated objects.
-    //http://stackoverflow.com/questions/3923826/nsfetchedresultscontroller-with-predicate-ignores-changes-merged-from-different
+    [self obtainPermanentIDsForInsertionsInContext:context];
+    [self saveContextToRoot:context andWait:YES];
+}
+
+/**
+ Save a managed object context and all of its parent contexts.  
+ This method is called recursively on the context's parentContext.
+ 
+ @param context The context to save
+ @param wait Whether to save synchronously (YES) or asynchronously (NO)
+ */
+- (void)saveContextToRoot:(NSManagedObjectContext *)context
+                  andWait:(BOOL)wait
+{
+    void (^saveBlock)() = ^{
+        NSError *error;
+        if ([context save:&error]) {
+            if (context.parentContext) {
+                [self saveContextToRoot:context.parentContext andWait:wait];
+            }
+        } else {
+            VOK_CDLog(@"Unresolved error %@, %@", error, error.localizedDescription);
+        }
+    };
     
-    for (NSManagedObject *object in [[notification userInfo] objectForKey:NSUpdatedObjectsKey]) {
-        [[self.managedObjectContext objectWithID:[object objectID]] willAccessValueForKey:nil];
+    //the root context can be saved asynchronously
+    if (context == self.privateRootContext) {
+        wait = NO;
```

Comment:
This seems a bit weird - this is basically saying no matter what, don't wait if it's the root context. I'm not sure I'd expect that behavior as a caller.